### ISSUE TITLE
Missing_OAuth_Token - invalid_credentials

### ISSUE THREAD

**Issue by ghost** [richardvanhook/omniauth-salesforce#13]
ERROR -- omniauth: (salesforce) Authentication failure! invalid_credentials: OAuth2::Error, Missing_OAuth_Token

we started getting this error recently, any suggestions. It is an intermittent issue.

**Comment by dchaley** [comment]
FYI I got around this by changing a post to a get:
https://github.com/docsend/omniauth-salesforce/

and in particular,
https://github.com/docsend/omniauth-salesforce/commit/a3445d8734ebec7fa07b6f49ceae23c47b7066da

I'm not sure if it's relevant to your issue, but it worked for me. Full disclosure: I didn't track down _why_ changing to a get works. I just observed that it does... so, YMMV!

(BTW, this is out of date, the newly maintained version is: https://github.com/realdoug/omniauth-salesforce)

**Comment by ghost** [comment]
Cool,will give it a try, thanks for the response.

**Comment by msreekm** [comment]
Thanks dchaley ,that worked.

I am wondering how changing from POST to GET fixed the issue!

**Comment by dchaley** [comment]
My suspicion is that it's a Faraday version issue, because the code works as-is in the sample (which uses Faraday 0.7.6). More info in my pull request here: https://github.com/realdoug/omniauth-salesforce/pull/12 (<-- that is the currently maintained site for this project, afaict)

I think it's that the POST requires the oauth_token to be part of a multipart form body, whereas the GET puts it as a parameter; I am guessing that the older version of Faraday did the right thing but the newer one is being stricter or an interface changed, or something like that.

What version of Faraday are you using @msreekm ?

**Comment by msreekm** [comment]
faraday (0.8.9)  and it was working fine with version 0.8.8.

**Comment by GLandes** [comment]
Hi. I am experiencing same issue. Tried to run your application "in action" getting
An error occurred in the application and your page could not be served. Please try again in a few moments.
If you are the application owner, check your logs for details.

Thanks. 
--GL

**Comment by sb8244** [comment]
Follow up that this is fixed with the latest version at realdoug.

**Comment by msreekm** [comment]
Hi sb8244- just curious, was this fixed recently @realdoug and do you know what was causing it to fail?

**Comment by sb8244** [comment]
@msreekm yea, I fixed it. The issue (I think) is that the old way was to POST the access_token in the body of the request. This may have worked in the past, but I think that SFDC changed their encoding so that spaces in the post body trip it up. I came to this conclusion while talking to Pat Patterson at Dreamforce and he said that is a likely cause (although blamed the library and not SFDC).

In any case, the proper way of sending an access_token is _always_ in the header and never in the url or POST body.

**Comment by kenleezle** [comment]
I had the same issue and changing the post to get fixed the issue for me.

**Comment by sb8244** [comment]
@kenleezle pull in the latest omniauth-salesforce and it's fixed through the header method.

**Comment by realdoug** [comment]
Correct.  Also, the source code for the gem is located here: https://github.com/realdoug/omniauth-salesforce so any issues, please open them under that repo.
